pageserver/src/tenant_threads.rs

```diff
@@ -1,77 +1,173 @@
 //! This module contains functions to serve per-tenant background processes,
 //! such as compaction and GC
+
 use crate::repository::Repository;
-use crate::tenant_mgr;
 use crate::tenant_mgr::TenantState;
+use crate::thread_mgr::ThreadKind;
+use crate::{tenant_mgr, thread_mgr};
 use anyhow::Result;
-use std::time::Duration;
+use once_cell::sync::OnceCell;
+use tokio::sync::mpsc::{self, Sender};
 use tracing::*;
 use utils::zid::ZTenantId;
 
 ///
-/// Compaction thread's main loop
+/// Compaction task's main loop
 ///
-pub fn compact_loop(tenantid: ZTenantId) -> Result<()> {
-    if let Err(err) = compact_loop_ext(tenantid) {
+async fn compaction_loop(tenantid: ZTenantId) -> Result<()> {
+    if let Err(err) = compaction_loop_ext(tenantid).await {
         error!("compact loop terminated with error: {:?}", err);
         Err(err)
     } else {
         Ok(())
     }
 }
 
-fn compact_loop_ext(tenantid: ZTenantId) -> Result<()> {
+async fn compaction_loop_ext(tenantid: ZTenantId) -> Result<()> {
     loop {
         if tenant_mgr::get_tenant_state(tenantid) != Some(TenantState::Active) {
             break;
         }
         let repo = tenant_mgr::get_repository_for_tenant(tenantid)?;
         let compaction_period = repo.get_compaction_period();
 
-        std::thread::sleep(compaction_period);
-        trace!("compaction thread for tenant {} waking up", tenantid);
+        tokio::time::sleep(compaction_period).await;
+        trace!("compaction loop for tenant {} waking up", tenantid);
 
         // Compact timelines
         let repo = tenant_mgr::get_repository_for_tenant(tenantid)?;
-        repo.compaction_iteration()?;
+        let compaction_result =
+            tokio::task::spawn_blocking(move || repo.compaction_iteration()).await;
+        match compaction_result {
+            Ok(Ok(())) => {
+                // success, do nothing
+            }
+            Ok(Err(e)) => {
+                anyhow::bail!(e.context("Compaction failed"));
+            }
+            Err(e) => {
+                anyhow::bail!("Compaction join error {}", e);
+            }
+        }
     }
 
     trace!(
-        "compaction thread stopped for tenant {} state is {:?}",
+        "compaction loop stopped for tenant {} state is {:?}",
         tenantid,
         tenant_mgr::get_tenant_state(tenantid)
     );
     Ok(())
 }
 
+static START_GC_LOOP: OnceCell<Sender<ZTenantId>> = OnceCell::new();
+static START_COMPACTION_LOOP: OnceCell<Sender<ZTenantId>> = OnceCell::new();
+
+/// Spawn a task that will periodically schedule garbage collection until
+/// the tenant becomes inactive. This should be called on tenant
+/// activation.
+pub fn start_gc_loop(tenantid: ZTenantId) -> Result<()> {
```

Comment:
> Result

but there are no result handling inside, only `unwrap`s. 
I think we should leave either, but not both. I'd go with `anyhow::Result` as more safe approach and since its `.context` works on `Option` just as well.


----

As a NOT style note, I find `-> anyhow::Result<()>` more readable since there's already an `std::result::Result` implicitly imported with the std prelude.

Reply:
Same for the compaction loop.